### ISSUE TITLE
using decision variable as indices

### ISSUE THREAD

**Issue by SHosseini94** [IBMDecisionOptimization/docplex-examples#43]
Hi, 
I am modeling a vehicle routing and scheduling problem using CP. I have 2 kinds of decision variables, that is, sequencing and start time of each node. The sequencing variable contains the node number, and start time of each node depends on the order it is traveled. Thus, I require to use sequencing variable as an index of start time variables. But I got this error: <docplex.cp.expression.CpoIntVar object at 0x0000022A9E070C70>

It is how I write my code: x={d:sub.integer_var_list(n[d],0, len(I)-1,"X") for d in D} start={d:{i:sub.interval_var(start=[t[0][i],T-p[i,d]-t[i][len(I)-1]],size=p[i,d]) for i in C[d]} for d in D}

sub.add(sub.sum(sub.end_of(start[d][x[d][n[d]-2]]),t[x[d][n[d]-2]][x[d][n[d]-1]])<=T)

I appreciate you to help me in this regard

**Comment by ooudot** [comment]
Hello,

Can you please be more precise and give:
- the complete stack that is displayed when the error occured,
- if possible, your source code to allow us to reproduce

**Comment by SHosseini94** [comment]
Thanks for response 
Here is the complete error:

<ipython-input-11-831280d9449f> in subproblem(k, C, D, I, p, t, T, L)
     33         sub.add(sub.element(x[d],n[d]-1)==len(I)-1)
     34         sub.add(sub.element(start[d],0)==0)                # the start time of depot is 0
---> 35         sub.add(sub.sum(sub.end_of(start[d][x[d][n[d]-2]]),t[x[d][n[d]-2]][x[d][n[d]-1]])<=T) # time limit
     36         for v in range(n[d]-1):
     37             sub.add(sub.end_at_start(start[d][x[d][v]],start[d][x[d][v+1]],t[x[d][v]][x[d][v+1]]))   #travel time between nodes

KeyError: <docplex.cp.expression.CpoIntVar object at 0x00000283C7FB4900>


There is my source code link:
https://github.com/SHosseini94/routing-shceduling-cp/blob/main/my_cp.ipynb

**Comment by ooudot** [comment]
Hello,

Your problem does not seems to concern docplex itself. You are using the "start" dictionnary that you have initialized with integer keys and that you access later with integer variable object. 
This explains the error "KeyError: <docplex.cp.expression.CpoIntVar object at 0x00000283C7FB4900>".

**Comment by SHosseini94** [comment]
you are true, but when I changed the "start" to initialize with integer variable object in this way:
start={d:{i:sub.interval_var(start=[t[0][i],T-p[i,d]-t[i][len(I)-1]],size=p[i,d]) for i in **x[d]**} for d in D} 

I get this error:
>>>TypeError: list indices must be integers or slices, not CpoIntVar

Is there any function, method in docplex.cp to let me define a variable index for another variable? Because using variable indices 
is common in CP Modeling

**Comment by ooudot** [comment]
You are mixing modeling and solving phases.
If you want, during solve, to access an array using an index that is in a variable, you have to use the element constraint (https://rawgit.com/IBMDecisionOptimization/docplex-doc/master/docs/cp/docplex.cp.modeler.py.html#docplex.cp.modeler.element)

**Comment by SHosseini94** [comment]
I do this and use element int this way:
sub.add(sub.less_or_equal(sub.sum(sub.end_of(start[d][sub.element(x[d],n[d]-2)]),t[sub.element(x[d],n[d]-2)][sub.element(x[d],n[d]-1)]),T))

But I got another error:
KeyError: <docplex.cp.expression.CpoFunctionCall object at 0x0000029E94ACAA40>

![Capture 1JPG](https://user-images.githubusercontent.com/73197807/111674594-8ea5d480-87f2-11eb-99ae-cf07ee7fea96.JPG)

**Comment by ooudot** [comment]
KeyError is a Python exception saying that you are accessing a dictionary with a key that is not inside. This is not a problem related to docplex.cp itself.
As your line is complex and full of indexed accesses, I recommand that you split it in multiple unitary expressions to precisely determine where the problem comes from.

**Comment by SHosseini94** [comment]
Hi,
I have 2d array of integers, t. I want to retrieve its t[i][j] element by using **element** function. How can I call this function when 
both i and j are the value of an integer variable x; t[x[1]][x[2]]???
I appreciate you guide me in this regard

Thanks

**Comment by ooudot** [comment]
Hi,

There is no 2D arrays in the modeling language. You have to convert it in a single dimension array.

For example, if you have an array of width W and height H, you can create a single dimension array as the concatenation of all rows. If r is row index and c column index, you can access it using t[r * W + c] instead of t[r][c].

Constraint element has to be used in the same way. If x and y are the variables containing your indices r and c, expression to access element t[r][c] is element(t, r * W + c).

Hope this helps.
